cloudinit/sources/DataSourceAkamai.py

```diff
@@ -0,0 +1,372 @@
+
+            # fetch general metadata
+            metadata = url_helper.readurl(
+                self._build_url("metadata", use_v6=use_v6),
+                timeout=30,
+                sec_between=2,
+                retries=2,
+                headers={
+                    "Accept": "application/json",
+                    "Metadata-Token": token,
+                },
+            )
+            self.metadata = json.loads(str(metadata))
+
+            # fetch user data
+            userdata = url_helper.readurl(
+                self._build_url("userdata", use_v6=use_v6),
+                timeout=30,
+                sec_between=2,
+                retries=2,
+                headers={
+                    "Metadata-Token": token,
+                },
+            )
+            self.userdata_raw = str(userdata)  # type: ignore
+            try:
+                self.userdata_raw = b64decode(self.userdata_raw).decode()  # type: ignore
+            except binascii.Error as e:
+                LOG.warning(f"Failed to base64 decode userdata due to {e}")
+        except url_helper.UrlError as e:
+            # we failed to retrieve data with an exception; log the error and return
+            # false, indicating that we should retry using a different network if
+            # possible
+            LOG.warning(
+                f"Failed to retrieve metadata using IPv{'6' if use_v6 else '4'} due to {e}"
+            )
+            return False
+
+        return True
+
+    def _get_data(self) -> bool:
+        """
+        Overrides _get_data in the DataSource class to actually retrieve data
+        """
+        LOG.debug("Getting data from Akamai DataSource")
+
+        if not is_on_akamai():
+            LOG.info("Not running on Akamai, not running.")
+            return False
+
+        local_instance_id = get_local_instance_id()
+        self.metadata = {
+            "instance-id": local_instance_id,
+        }
+        availability = self._should_fetch_data()
+
+        if availability != MetadataAvailabilityResult.available:
+            if availability == MetadataAvailabilityResult.not_available:
+                LOG.info(
+                    "Metadata is not available, returning local data only."
+                )
+                return True
+
+            LOG.info(
+                "Configured not to fetch data at this stage; waiting for "
+                "a later stage."
+            )
+            return False
+
+        network_context_managers = self._get_network_context_managers()
+        for manager, use_v6 in network_context_managers:
+            with manager as m:
+                done = self._make_requests(use_v6=use_v6)
+                if done:
+                    # fix up some field names
+                    self.metadata["instance-id"] = self.metadata.get(
+                        "id",
+                        local_instance_id,
+                    )
+                    break
```

Comment:
If the loop does not break, then the ds has not been update. Should we return false in that case?

```suggestion
                    break
        else:
            return False
```

Reply:
I tested this, but I found it didn't behave as expected.

If the metadata service was down when cloud-init ran and DataSourceAkamai returned `False`, cloud-init moved onto DataSourceNone and populated a cached instance-id of `iid-datasource-none`.  This was fine, except that on the next boot cloud-init would ask DataSourceAkamai for the instance id to check that the cache was valid, and our code would return the actual instance id (taken from the `system-serial-number` dmi field).  This, of course, did not match, invalidating the cache and causing cloud-init to reprovision the instance, generating new host keys and possibly applying any user data that was sent up with the instance.

My concern with this behavior is that a user may create an instance without user data during an outage of our metadata service, not notice that DataSourceNone was used, and then at some point in the future reboot and unexpectedly see their host keys change, which would be alarming.

This DataSource will set up metadata [here](https://github.com/canonical/cloud-init/pull/4167/files#diff-20689f9a33920f781eeb32d8802f9f931443049d9dff5c87151f3df00fb3ccadR304-R306) that is used if it's either configured not to run, or if the service is unavailable - if a user really needs cloud-init to re-run, cleaning and rebooting is a simple way to do that without the risk of it happening unexpectedly.

Reply:
Sounds good. Do you think it is worth to leave some kind of breadcrumb in the log for the users to know this happened? Something like:

```python
    else:
        LOG.warning("Unable to reach IMDS.")
```

Reply:
Yep, that makes sense - I'll push that up alongside the changes to satisfy the linters 👍